Wire in signing certs taken from IDM

tvs opened the pull request.

tvs committed:
Added configuration to IdentityZone provisioning

Includes the signing keys and access/refresh token validity

tvs committed:
Added log4j 2.2 to dependency list

Cleaned up a few lingering loggers that went unused

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
+        IdentityZone zone = null;
+        String systemTenant = this._idmClient.getSystemTenant();
+        
+        if (tenantName.equals(systemTenant)) {
```
equalsignorecase ...

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
+        IdentityZone zone = null;
+        String systemTenant = this._idmClient.getSystemTenant();
+        
+        if (tenantName.equals(systemTenant)) {
```
equalsIgnorecase... or you can add isSystemTenant util function to vmidentity utils.

tvs replied:
Good catch, thanks.

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
         }
     }
+    
+    private IdentityZone identityZoneForTenant(String tenantName) throws Exception {
```
let's rename this, so it is obvious this should only be called for non-system tenant.

tvs replied:
I'm not entirely sure that it should only be for the non-system tenant. It's currently called only for the non-system tenant, but I think we'll wind up wanting to call it for both once the configuration is further fleshed out.

mpotapova-vmw replied:
it should be for untill we re-wire on higher level.... but we can keep the function name as is.

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
+    }
+    
+    private void configureIdentityZone(String tenantName, IdentityZone zone) throws Exception {
+        IdentityZoneConfiguration config = this._configs.get(tenantName);
```
with configs wired to idm, we should not longer need to have in-memory configs ... ?

tvs replied:
They're not fully wired - updates are unavailable, so at least some of it still get stored only in memory.

mpotapova-vmw replied:
we should fully wire this to idm. we can defer the write/change part, and if somebody needs to change to do it via LW. originally we have to keep in-memory, to allow modification done during bootstrap. now that we won;t need bootstrap, we can remve it. this removal can be separate checkin.

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -12,23 +12,25 @@
-import org.cloudfoundry.identity.uaa.provider.VmidentityIdentityProviderProvisioning;
 import org.cloudfoundry.identity.uaa.util.VmidentityUtils;
 
+import com.google.common.primitives.Ints;
```
is this something already used by rest of code? otherwise, can we make it without this?

tvs replied:
Guava is already included in the dependency list - yes.

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
+        String tenantName = id;
+        IdentityZone zone = null;
+        
+        if (id.equalsIgnoreCase(IdentityZone.getUaa().getId())) {
```
can we please keep VmidentityUtils.getTenantName ?

tvs replied:
The trouble is that this structure is being used to do two different things: not only are we retrieving the true system tenant name (based on the fact we were in the default UAA zone), but we're fetching the default `IdentityZone` as well.

To use that method, I have to do two things: I have to fetch the system tenant's name (which is kind of pointless when we're not in the default zone) and I have to then compare the output of `#getTenantName()` against the system tenant to determine which zone I pull. For example:

    String systemTenant = this._idmClient.getSystemTenant();
    String tenantName = VmidentityUtils.getTenantName(id, systemTenant);
    if (tenantName.equalsIgnoreCase(systemTenant)) {
         zone = IdentityZone.getUaa();
    } else {
        zone = identityZoneForTenant(tenantName);
    }

Doing that both creates an unnecessary remote IDM call in the non-UAA zone (e.g. a named tenant) and creates a second comparison that's functionally the same as comparing the zone against the default UAA zone. I also can't remove the remote IDM call at this point, because I'll wind up needing the name of the system tenant when it comes time to fetch the `IdentityZone`.

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
+        
+        Map<String, String> keys = new HashMap<String, String>();
+        PrivateKey privateKey = _idmClient.getTenantPrivateKey(tenantName);
+        StringWriter writer = new StringWriter();
```
let's handle all closeables with try-finally, or try-with-resources, which is handy...

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -79,42 +79,84 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
+        config.getTokenPolicy().setAccessTokenValidity(
+                Ints.saturatedCast(_idmClient.getMaximumBearerTokenLifetime(tenantName)));
+        config.getTokenPolicy().setRefreshTokenValidity(
+                Ints.saturatedCast(_idmClient.getMaximumBearerRefreshTokenLifetime(tenantName)));
```
q: isn;t there also saml config, which also has signers ?

tvs replied:
There is - I have the change for that sitting here, but haven't finished it just yet. It seems to also want a private key password (`SamlConfig#setPrivateKeyPassword(String)`) and possibly a separate certificate. I'm trying to suss out whether we need the password (I don't think we do). And what the certificate is for.

mpotapova-vmw reviewed (commented):
thanks. couple questions below.

tvs committed:
Added privateKey and certificate to SamlConfiguration.

Fixed a couple of issues with the try-catch block and made a tenant
comparison case insensitive.

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/zone/VmidentityIdentityZoneProvisioning.java:
```diff
@@ -79,42 +80,95 @@ public IdentityZone retrieveBySubdomain(String subdomain) {
+        PrivateKey privateKey = _idmClient.getTenantPrivateKey(tenantName);
+        StringWriter writer = new StringWriter();
+        
+        try (PEMWriter pemWriter = new PEMWriter(writer)) {
```
i would do 
try(StringWriter writer = new StringWriter(), PEMWriter pemWriter = ...) {}

tvs committed:
Fixed issues with SAML configuration.

Forced all private key/certificate serialization to use the RSA
provider.

tvs committed:
Plumbed through URIs from IDM

tvs committed:
Reverted a few superfluous changes

mpotapova-vmw commented on uaa/src/main/webapp/WEB-INF/spring/saml-providers.xml:
```diff
@@ -41,6 +41,14 @@
             </security:filter-chain-map>
         </bean>
 
+        <bean id="idmClient" class="com.vmware.identity.idm.client.CasIdmClient">
```
we already have idm client somewhere, can we keep that single copy ? or dis this just move from there?

mpotapova-vmw commented on uaa/src/main/webapp/WEB-INF/spring/saml-idp.xml:
```diff
@@ -29,6 +29,14 @@
     <bean id="samlIdpAuthenticationProvider"
         class="org.cloudfoundry.identity.uaa.provider.saml.idp.IdpSamlAuthenticationProvider" />
 
+    <bean id="idmClient" class="com.vmware.identity.idm.client.CasIdmClient">
```
all these xmls are actually fragments stiched together. can we keep single idmClient bean and single uaaUrlRetriever, and make use of them both in all these places needed? thanks.

mpotapova-vmw commented on server/src/main/java/org/cloudfoundry/identity/uaa/util/VmidentityUaaURLRetriever.java:
```diff
@@ -0,0 +1,49 @@
+
+        // Need to strip off any websso path that lingers and add the uaa path instead
+        URL url = new URL(serviceUri);
+        builder.append(url.getProtocol() + "://");
```
why do we need ips? if so, do we need ipv6s ?

mpotapova-vmw reviewed (commented):
overall looks good. let;'s please consolidate idmClient & urlretriver and submit. thanks!

tvs committed:
Consolidated idmClient and uaaUrlRetriever into a single instance
